### PR TITLE
Immutable payables

### LINKED ISSUE #1766: Immutable payables
### Describe the change
A common problem, also in our current codebase, is that we often want to prevent changing payables when they are already paid. Think of Food/Sales orders. Right now there is no good way to fix this fundamentally. Sales orders only prevent saving when the amount of the payment and the order don't match (but doesn't update the description, for example). Food orders since #1765 only does the check in the view.

A way to solve this fundamentally would be the following:
- Payables get an additional boolean property `immutable_after_payment` that defaults to False, but can be enabled for models where we want to apply this
- The payments app implements a pre_save_signal on payable models that raises an exception if a model that is `immutable_after_payment` and has a payment is tried to be saved. (Or it tries to delete the payment and will raise the exception when that is not possible anymore - for example because the payment has already been processed).

### Motivation
This is a problem that applies to all payables and could be a generic fix.

### Current implementation
Messy, not really 

### Suggested implementation
See above

### PR DESCRIPTION
**Pull request by JobDoesburg**
Closes #1766 

### Summary
Connects a model pre_save signal when registering payables. The signal will prevent saving models with certain fields changed if the payable is actually paid. This gives a hard guarantee that payable models are immutable after payment, integrated in the payments app.

### How to test
1. Try to save payable models or fields with a foreign key to payable models, that are marked as immutable, when the payable is paid.
2. Be careful for edge cases: creating new objects that directly have a payment too, what if we try to delete a payment from a model, etc.

### CODE REVIEW

File: website/payments/payables.py [hunk 1]
```diff
@@ -66,6 +91,75 @@ def get_payable(self, model: Model) -> Payable:
...
+        else payable.immutable_model_fields_after_payment
+    )
+    for field in immutable_fields:
+        if getattr(old_instance, field) != getattr(instance, field):
+            raise PaymentError("Cannot change this model")
+
+
+def prevent_saving_related(foreign_key_field):
+    def prevent_related_saving_paid_after_immutable(sender, instance, **kwargs):
+        nonlocal foreign_key_field
```
Comment (pingiun) [inline website/payments/payables.py:141]:
```suggestion
```

**Review by pingiun — APPROVED** [review]

**Review by DeD1rk — APPROVED** [review]
Looks good! I do think it good use some documentation.

Something to keep in mind: 

> gives a hard guarantee

This is only a guarantee when `save()` is always used. Bulk updates and deletes can still violate it.

**Review by se-bastiaan — COMMENTED** [review]
Looks nice, just one question of something I don't entirely know.

File: website/sales/payables.py [hunk 2]
```diff
@@ -25,6 +27,29 @@ def can_manage_payment(self, member):
...
+        return True
+
+    @classproperty
+    def immutable_foreign_key_models(self):
+        return {OrderItem: "order"}
+
+    @classproperty
+    def immutable_model_fields_after_payment(self):
+        return {
+            Order: [
```
[the remark this answers is not in the document]
Comment (JobDoesburg) [reply website/sales/payables.py:41]: Good point. What I found online is that it uses `__hash__` and `__eq__` in the background and that it should work as long as you don't override those... Class name as string would work I think, but in the code I would then need to retrieve the class again from the string to register the hook, so I think this is in fact simpler

[the remark this answers is not in the document]
Comment (JobDoesburg) [reply website/sales/payables.py:41]: @se-bastiaan do you want me to investigate changing this? Or is it fine to keep it this way

**Review by se-bastiaan — APPROVED** [review]

**Review by se-bastiaan — APPROVED** [review]
